### DISCUSSION TITLE
Generazione Token

### DISCUSSION THREAD

**Discussion by elsad3631** [fattureincloud/api#155] [Q&A] [answered]
Salve.
Sto implementando il codice per la generazione del token usando il vostro sdk c# ma la richiesta mi restituisce una bad request.

                string clientId = _configuration.GetValue<string>("CloudInvoice:ClientId");
                string clientSecret = _configuration.GetValue<string>("CloudInvoice:ClientSecret");
                string redirectUri = _configuration.GetValue<string>("CloudInvoice:RedirectUri");
                OAuth2AuthorizationCodeManager oauth = new OAuth2AuthorizationCodeManager(clientId, clientSecret, redirectUri);

                var scopes = new List<Scope>() {
                    Scope.SETTINGS_ALL,
                    Scope.ISSUED_DOCUMENTS_INVOICES_READ
                };
                var url = oauth.GetAuthorizationUrl(scopes, "EXAMPLE_STATE");

                var param = oauth.GetParamsFromUrl(url);
                var code = param.AuthorizationCode;
                var state = param.State;

                var tokenObj = oauth.FetchToken(code);
                var accessToken = tokenObj.AccessToken;
                var refreshToken = tokenObj.RefreshToken;

Questo è il codice per la generazione del token. 
Il metodo GetParamsFromUrl mi restituisce param.AuthorizationCode con valore null.
Per questo motivo il metodo FetchToken mi restituisce un errore.
Il clientId e clientSecret li ho presi dall'app che ho creato.
Come RedirectUri ho inserito (https://www.yourapplication.com/redirect) come da documentazione.
Ho provato anche ad inserire https://localhost:7099/oauth ma non funziona ugualmente.

**Comment by valmoz** [comment, marked as answer]
Buongiorno.
Prima di tutto le consiglio di controllare tutti i metodi di autenticazione disponibili, per verificare se effettivamente OAuth è il metodo migliore per lei, ciò dipende dal suo caso d'uso e dalle risorse a sua disposizione. [Authentication](https://developers.fattureincloud.it/docs/basics/authentication)

Ad esempio, nel caso in cui lei dovesse usare la sua applicazione solo con il suo account Fatture in Cloud (e non con quello di altre aziende sconosciute a priori) potrebbe essere più comodo utilizzare il manual token.

Per poter utilizzare OAuth2 Authorization Code Flow, esistono due prerequisiti:

1. poter redirigere l'utente su una pagina web tramite browser
2. poter esporre un endpoint lato backend per ottenere la risposta dal redirect

Se questi due prerequisiti non sono applicabili nel suo caso, consiglio di passare ad un altro metodo di autenticazione.

Passiamo ora al codice che ha inviato. 
Prima di tutto, il Redirect URI non può essere l'url fasullo https://www.yourapplication.com/redirect (che abbiamo inserito solo come esempio), ma dev'essere un endpoint vero esposto dalla sua applicazione. In fase di test è eventualmente possibile usare localhost, come da lei indicato.

Ulteriore problema: da quello che vedo lei non sta svolgendo il processo di redirect, per questo sta ricevendo un errore.

Una volta ottenuto l'[Authorization Url](https://developers.fattureincloud.it/docs/authentication/code-flow/sdk-oauth-helper#2%EF%B8%8F%E2%83%A3-redirect-the-user-to-the-fatture-in-cloud-authorization-page), lei deve reindirizzare l'utente alla pagina indicata dall'url stesso. Ciò significa che l'utente deve poter accedere alla pagina ed effettuare l'autenticazione.

Una volta effettuato questo, la nostra pagina redirigerà l'utente sempre tramite browser  all'endpoint da lei indicato (quindi il redirect Uri) effettuando una GET, e lei dovrà estrarre i [parametri indicati da query string](https://developers.fattureincloud.it/docs/authentication/code-flow/sdk-oauth-helper#3%EF%B8%8F%E2%83%A3-obtain-the-authorization-code). Le faccio notare che la nostra funzione getParamsFromUrl serve solo nel caso in cui il framework rest che state utilizzando non sia in grado di recuperare in autonomia i path params, cosa molto improbabile.

Per comodità, per controllare di aver ottenuto effettivamente un parametro valido vi consiglio anche di controllare che l'authorization  code inizi con il prefisso "a/"

Solo a quel punto potrà effettuare la chiamata a Fetch token.

Resto a disposizione per ulteriori domande
Mauro

**Comment by elsad3631** [comment]
OK grazie mille ora funziona!
